### PR TITLE
Add support for customizing exported metrics

### PR DESCRIPTION
**Pull request by dinooliva**
Adds 2 features for customizing the exported metric name (i.e. the name used in the generated metric descriptor):

1. A mapping from the Prometheus metric name to a new name. This feature provides compatibility when the Prometheus metric name doesn't map directly onto the exported metric name (e.g. providing backward compatibility when the original metric name is being changed).

2. A per-job metrics prefix. This feature allows users to specify a metric prefix for each job in the scrape configuration, which is required for systems like e.g. Kubernetes.

### CODE REVIEW

File: receiver/README.md [hunk 1]
```diff
@@ -151,22 +151,42 @@ receivers:
...
 
 #### Example
 ```yaml
 receivers:
     prometheus:
-      include_filter: {
-        "localhost:9777" : [http/server/server_latency, custom_metric1],
-        "localhost:9778" : [http/client/roundtrip_latency],                
+      job_export_prefix: {
+        "localhost:9777" : "opencensus.io/prometheus"
```
Comment (bogdandrutu) [inline receiver/README.md]: If this is per target as well as the filters, why do we not extend Target to have more properties:
```
    prometheus:
      config:
        scrape_configs:
          - job_name: 'opencensus_service'
            scrape_interval: 5s
            static_configs:
              - target: 'localhost:8889'
              include_filter: [http/server/server_latency, custom_metric1]
              export_prefix: "opencensus.io/prometheus"
          - job_name: 'jdbc_apps'
            scrape_interval: 3s
            static_configs:
              - target: 'localhost:9777'
```

Comment (dinooliva) [reply receiver/README.md]: The config portion of the specification comes from Prometheus - I don't think there is a simple way to extend it.

Comment (bogdandrutu) [reply receiver/README.md]: Ok, let me first understand if the mapping I suggested makes sense. I mean if the prefix apply to the target inside a job not to the whole job.

File: receiver/README.md [hunk 2]
```diff
@@ -151,22 +151,42 @@ receivers:
               - targets: ['localhost:9777']
 ```
 
-### Include Filter
-Include Filter provides ability to filter scraping metrics per target. If a filter is specified for
-a target then only those metrics which exactly matches one of the metrics specified in the `Include Filter` list will be scraped.
-Rest of the metrics from the targets will be dropped.
+### Job Export Prefix
+Job Export Prefix provides the ability to add a metrics prefix per target. If a prefix is specified for a target then
+those metrics are scraped from that target will have the prefix added in the exported name.
```
Comment (songy23) [inline receiver/README.md]:
```suggestion
those metrics scraped from that target will have the prefix added in the exported name.
```

File: receiver/README.md [hunk 3]
```diff
@@ -151,22 +151,42 @@ receivers:
...
 
 #### Syntax
 - Endpoint should be double quoted.
-- Metrics should be specified in form of a list.
+- Prefix should be double quoted with a "/" at the end.
```
Comment (songy23) [inline receiver/README.md]: Looks like none of the examples below has a "/" at the end.

File: receiver/README.md [hunk 4]
```diff
@@ -151,22 +151,42 @@ receivers:
...
+        "localhost:9777" : "opencensus.io/prometheus"
+        "localhost:9778" : "opentelemetry.io/prometheus"
+      }
+      config:
+        scrape_configs:
+          ...
+```
+
+### Metric Export Name
+Metric Export Name provides the ability to transform a given Prometheus metric name into an new name. If an export name is specified then
```
Comment (songy23) [inline receiver/README.md]:
```suggestion
Metric Export Name provides the ability to transform a given Prometheus metric name into a new name. If an export name is specified then
```

File: receiver/README.md (hunk 2 above)
Comment (dinooliva) [reply receiver/README.md]: Done.

File: receiver/README.md (hunk 3 above)
Comment (dinooliva) [reply receiver/README.md]: Done.

File: receiver/prometheusreceiver/internal/metricsbuilder.go [hunk 5]
```diff
@@ -70,7 +74,6 @@ func (b *metricBuilder) AddDataPoint(ls labels.Labels, t int64, v float64) error
 		b.hasInternalMetric = true
 		lm := ls.Map()
 		delete(lm, model.MetricNameLabel)
-		b.logger.Infow("skip internal metric", "name", metricName, "ts", t, "value", v, "labels", lm)
```
Comment (songy23) [inline receiver/prometheusreceiver/internal/metricsbuilder.go]: Why is this removed?

File: receiver/prometheusreceiver/internal/transaction.go [hunk 6]
```diff
@@ -142,6 +151,13 @@ func (tr *transaction) Commit() error {
 		if tr.jobsMap != nil {
 			metrics = NewMetricsAdjuster(tr.jobsMap.get(tr.job, tr.instance), tr.logger).AdjustMetrics(metrics)
 		}
+
+		res := "*** Metrics: "
+		for _, metric := range metrics {
+			res += metric.MetricDescriptor.Name + ", "
+		}
+		tr.logger.Info(res)
```
Comment (songy23) [inline receiver/prometheusreceiver/internal/transaction.go]: This looks like a debugging log.

File: receiver/prometheusreceiver/internal/metricsbuilder.go (hunk 5 above)
Comment (dinooliva) [reply receiver/prometheusreceiver/internal/metricsbuilder.go]: It spams the logger every time a job is scraped - the internal metrics are expected and skipping them is the expected behavior so there's no need to log the event.

File: receiver/prometheusreceiver/internal/transaction.go (hunk 6 above)
Comment (dinooliva) [reply receiver/prometheusreceiver/internal/transaction.go]: Done.

File: receiver/prometheusreceiver/internal/metricfamily.go [hunk 7]
```diff
@@ -75,6 +76,17 @@ func newMetricFamily(metricName string, mc MetadataCache) MetricFamily {
 	}
 }
 
+func exportName(metricName string, metricExportNameMap map[string]string, exportPrefix string) string {
```
Comment (tigrannajaryan) [inline receiver/prometheusreceiver/internal/metricfamily.go]: Can you please add a comment to explain the purpose of this function?

File: receiver/prometheusreceiver/internal/metricsbuilder.go (hunk 5 above)
Comment (tigrannajaryan) [inline receiver/prometheusreceiver/internal/metricsbuilder.go]: Why is this deleted?

Comment (songy23) [reply receiver/prometheusreceiver/internal/metricsbuilder.go]: I asked the same question a moment ago :) https://github.com/open-telemetry/opentelemetry-service/pull/317#discussion_r320469844

File: receiver/prometheusreceiver/internal/metricfamily.go (hunk 7 above)
Comment (dinooliva) [reply receiver/prometheusreceiver/internal/metricfamily.go]: Done.

File: receiver/prometheusreceiver/internal/metricsbuilder.go (hunk 5 above)
Comment (dinooliva) [reply receiver/prometheusreceiver/internal/metricsbuilder.go]: As I replied to @songy23: It spams the logger every time a job is scraped - the internal metrics are expected and skipping them is the expected behavior so there's no need to log the event.

I'm fine to keep it but it's not necessary.

**Comment by dinooliva** [comment]
@tigrannajaryan @songy23 - Updated the pr description and responded to comments - ptal

**Comment by dinooliva** [comment]
As discussed offline, we are going to provide this functionality as a processor. Thanks for all the help - they will be carried over to the subsequent code.
